NodalKernels

friedmud opened the pull request:
Still a few things to do, including making Jacobians work and putting in a lot more tests.  I also plan on making the threading actually work (currently, NodalKernels are only computed on one thread).  But to do that I need to add a few things to Assembly.

Putting this here to get it run through the testing system and get any feedback...

refs #3029

friedmud committed:
initial implementation for NodalKernels.  refs #3029
Still quite a lot to do, including making Jacobians work and putting in a lot more tests

friedmud committed:
switch to caching residual entries from nodal kernels and enable threading refs #3029

friedmud committed:
genericize jacobian caching so it can be used for nodal kernels as well. refs #3029

friedmud committed:
Add Jacobian contributions for NodalKernels refs #3029

friedmud commented:
This is close... but I need to make a `TimeNodalKernel` base class and a `NodalKernelTimeDerivative`... and verify that higher order time integration works with the new system.

It also needs a LOT more tests for all of the edge case (for instance, off-diagonal Jacobian entries aren't tested yet... and neither are many of the error conditions).

Of course... documentation also needs to be added......

friedmud committed:
restrict jac_test to 1 MPI refs #3029

friedmud committed:
add better time derivative system for NodalKernels refs #3029

friedmud committed:
Add a test and convergence script for time integration of NodalKernels refs #3029

friedmud commented:
Ok - This is ready to go as far as I can tell... BUT... @jwpeterson @andrsd there is _something_ up with time integration.  I added a test ( nodalkernels/high_order_time_integration ) that has convergence_study.py in it that will generate this plot:

![figure_1](https://cloud.githubusercontent.com/assets/870705/10569882/893ed880-75fa-11e5-8473-39ccbcdf3e32.png)

And... just... WTH.  The good news is that for implicit-euler and BDF2 the error DOES go down as it refines in time... BUT... you don't get better rates out of higher-order time integrators.  PLUS: Just what in the hell is up with crank-nocolson?

I'm sure that there is something I'm missing here.  I tried to be careful and get the correct residual contributions to go to the correct residual vector (time vs. non-time)... but I'm sure I messed something up somewhere.

Anyway... I think that this PR can still be merged for now and we can fix up this time integration thing later...

friedmud commented:
BTW: This _does_ still need more tests too... but it's going to be a little while before I have time to do that.  I'm smashed until Wednesday... wait... NEXT Wednesday (hopefully).

Sigh.

I _did_ add a ticket to keep track of this fact over here: #5826

friedmud commented:
I have no idea why that test isn't passing.  Do you guys?

It looks totally unrelated....

permcody commented:
Well we merged two large PRs just a little bit ago but they were in the TM
and PF modules respectively. Solid Mechanics should be working. You might
try a debug build locally?

On Mon, Oct 19, 2015 at 3:06 PM Derek Gaston <notifications@github.com>
wrote:

> I have no idea why that test isn't passing. Do you guys?
>
> It looks totally unrelated....
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5815#issuecomment-149347690>.
>

friedmud commented:
I'm pretty sure that these test failures are NOT associated with my stuff.  To me it looks like an actual bug in the Truss stuff.

Has anyone used that stuff lately?

friedmud commented:
There is definitely an actual bug here... I don't think this NodalKernel code is the _cause_ of it... but it looks like it poked the system in just the right way to uncover the bug.

I'm deep in debugging...

friedmud committed:
Why are you always incorrectly preparing? #3029

friedmud committed:
final cleanup on nodal kernels closes #3029

friedmud commented:
Ok - fixed the issue in b53956d (note the message ;-)

It was a small bug in `DisplacedProblem::reinitDirac()` completely unrelated to NodalKernels... but adding NodalKernels into the mix showed the issue.

I also added an extra commit to clean up a few things: mostly making sure that NodalKernel related activities aren't undertaken when you don't have any!

I think this should be good to go...

@andrsd @jwpeterson did you guys ever look at my time convergence plots above?  At some point we're going to need to figure out why NodalKernels don't do any better with high order time integration...

jwpeterson commented:
@friedmud sorry it took me a while to look at this, the number of lines changed was a bit daunting... However, I think the solution is rather simple -- you have to use a *fixed final time* while cutting the timestep (thus doing more and more timesteps to get to the final solution) when doing the convergence study.  So I changed this [one line](https://github.com/jwpeterson/moose/commit/fea9184a27a8a6a0bd75f7b18d382fded734c3e0) and ran a couple cases with BDF2:

```
../../../moose_test-opt -i high_order_time_integration.i Executioner/scheme=bdf2 Executioner/dt=1
../../../moose_test-opt -i high_order_time_integration.i Executioner/scheme=bdf2 Executioner/dt=.5
../../../moose_test-opt -i high_order_time_integration.i Executioner/scheme=bdf2 Executioner/dt=.25
../../../moose_test-opt -i high_order_time_integration.i Executioner/scheme=bdf2 Executioner/dt=.125
../../../moose_test-opt -i high_order_time_integration.i Executioner/scheme=bdf2 Executioner/dt=.0625
```

and got the following results
```
dt     Final error
1      4.090625e+03
.5     1.060039e+03
.25    2.699243e+02
.125   6.811050e+01
.0625  1.710725e+01
```
The slope of this data on the log-log plot is approximately 1.98.  You might also want to update your simple_plot.py to do fewer dt's as it now takes forever to run with the fixed end time ;-)

jwpeterson commented:
Unfortunately, this doesn't fix all the methods the way it fixed BDF2...

Crank-Nicolson doesn't converge at all...
```
dt     Final error
1      1.978900e+05
.5     2.093025e+05
.25    2.149056e+05
.125   2.176814e+05
.0625  2.190629e+05
```

... while Dirk is first-order
```
dt     Final error
1      4.510000e+04
.5     2.227500e+04
.25    1.106875e+04
.125   5.517188e+03
.0625  2.754297e+03
```

Crank-Nicolson does some sneaky stuff in `postSolve()` to shift the non-time residual forward in time so it can be reused at the next timestep.  Dirk calls `_fe_problem.getNonlinearSystem().sys().solve()` multiple times within a single timestep while manipulating _fe_problem.time().  Do you think either of those things would cause issues with NodalKernels?

friedmud commented:
Thanks for looking into this. I don't think either one should cause an
issue. I'm contributing to the very same residual vector that a regular
Kernel would. In fact... NodalKernels are even simpler because they are
pure ODEs in time... so they really should behave properly.

I'm sure it's just something simple I've overlooked.

BTW: I don't think this issue should hold up the PR. The solutions are
right for both BE and BDF2... which is good enough to release into the wild
while we work on higher order time integration.
On Tue, Oct 27, 2015 at 5:44 PM John W. Peterson <notifications@github.com>
wrote:

> Unfortunately, this doesn't fix all the methods the way it fixed BDF2...
>
> Crank-Nicolson doesn't converge at all...
>
> dt     Final error
> 1      1.978900e+05
> .5     2.093025e+05
> .25    2.149056e+05
> .125   2.176814e+05
> .0625  2.190629e+05
>
> ... while Dirk is first-order
>
> dt     Final error
> 1      4.510000e+04
> .5     2.227500e+04
> .25    1.106875e+04
> .125   5.517188e+03
> .0625  2.754297e+03
>
> Crank-Nicolson does some sneaky stuff in postSolve() to shift the
> non-time residual forward in time so it can be reused at the next timestep.
> Dirk calls _fe_problem.getNonlinearSystem().sys().solve() multiple times
> within a single timestep while manipulating _fe_problem.time(). Do you
> think either of those things would cause issues with NodalKernels?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5815#issuecomment-151654665>.
>

permcody commented on framework/src/base/ComputeNodalKernelsThread.C:
```diff
@@ -0,0 +1,85 @@
+/****************************************************************/
```
Definitely time to create ThreadedNodalLoop BASE CLASS

friedmud replied:
I can agree with that.  One thing I noticed during this work is that we still do NodalBC's _un_-threaded.  That should be fixable.  They will never be super efficient on threads... but they will be better than doing them serially like they are now.  I almost included that fix in this PR... but decided to get this in first and then come back and redo all of the nodal calculations with a threaded node loop base class.

Do you think it's needed for this PR?  Or can we merge it as is and fix this later.  If we can fix this later then I'll put in a ticket for it.

permcody replied:
I think we already have a ticket in there for that, we just need to dig it up and get to it.

permcody commented on framework/src/base/ComputeNodalKernelsThread.C:
```diff
@@ -0,0 +1,85 @@
+
+    num_cached++;
+
+    if (num_cached % 20 == 0) // Cache 20 nodes worth before adding into the residual
```
Not a big deal, but this should probably just be `>=` instead of `%` since you are resetting the number in the next statement.

friedmud replied:
That's a good point.  In ComputeResidualThread we don't reset that number... but I started to worry about overflow (this gets called a LOT... but I don't know about billions of times...).

I should probably change both of them...

permcody replied:
Not a deal breaker for this PR. If you get to it before I do, might as well make this a const class static too.

friedmud replied:
Ok - I'll fix this up in a later PR.

permcody commented:
:+1: with some caveats. I'd like to see a ticket for the higher order issue. @jwpeterson - if you think of anything else that needs a ticket, let's put one in. Otherwise, go ahead and merge.

friedmud commented:
Ok - I added a new Issue for Time Integration and added to the ThreadedNodeLoop ticket.  Unless someone else spots something here I think this is ready to go...

friedmud commented:
BTW: @permcody I think _this_ is the last System we'll ever need :wink:

permcody commented:
I'm sure it is...

permcody commented:
@jwpeterson is going to add his fix to your Python script, then we'll merge.

friedmud commented:
Sounds good. If he puts in a PR to me I'll merge it right away.

Derek
On Wed, Oct 28, 2015 at 11:14 AM Cody Permann <notifications@github.com>
wrote:

> @jwpeterson <https://github.com/jwpeterson> is going to add his fix to
> your Python script, then we'll merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/idaholab/moose/pull/5815#issuecomment-151877446>.
>
